File: Drum/sticking-engine-ts/src/rule26-runtime.ts
/**
 * rule26-runtime.ts — RULE 26: PERFORMANCE RUNTIME, PLAYBACK STATE &
 * DETERMINISTIC REPRODUCTION
 *
 * "Runtime faithfully executes solved performance; no hidden
 * re-sticking or random decision-making." This is purely a reader of an
 * already-solved performance -- it never calls back into Rules 1–25 to
 * re-decide anything.
 */

import type { FinalValidatedPerformance, PerformanceEvent } from './datamodel.js';

export class PerformanceRuntime {
  readonly performance: FinalValidatedPerformance;
  playheadS = 0.0;
  playing = false;
  private readonly sorted: readonly PerformanceEvent[];

  constructor(performance: FinalValidatedPerformance) {
    if (!performance.validation.approved) {
      throw new Error('cannot construct a runtime from an unapproved performance (Rule 25 gate)');
    }
    this.performance = performance;
    this.sorted = [...performance.events].sort((a, b) => a.timeSeconds - b.timeSeconds);
  }

  play(): void {
    this.playing = true;
  }

  pause(): void {
    this.playing = false;
  }

  seek(timeS: number): void {
    this.playheadS = Math.max(0.0, Math.min(this.performance.durationS, timeS));
  }

  /** Exactly the events due in a frame window. Same seed and version, same slice. */
  eventsBetween(t0: number, t1: number): PerformanceEvent[] {
    return this.sorted.filter((e) => t0 <= e.timeSeconds && e.timeSeconds < t1);
  }

  advance(dtS: number): PerformanceEvent[] {
    if (!this.playing) return [];
    const t0 = this.playheadS;
    const t1 = Math.min(this.performance.durationS, this.playheadS + dtS);
    const due = this.eventsBetween(t0, t1);
    this.playheadS = t1;
    return due;
  }

  exportSummary(): Record<string, number | string> {
    return {
      engineVersion: this.performance.engineVersion,
      seed: this.performance.seed,
      durationS: this.performance.durationS,
      eventCount: this.performance.events.length,
      validationIssues: this.performance.validation.issues.length,
    };
  }
}

File: Drum/sticking-engine/sticking_engine/rule26_runtime.py
"""
rule26_runtime.py — RULE 26: PERFORMANCE RUNTIME, PLAYBACK STATE &
DETERMINISTIC REPRODUCTION

Purpose: Execute approved performance data with a single authoritative clock
and reproducible state.

Core Principle: "Runtime faithfully executes solved performance; no hidden
re-sticking or random decision-making."

Implementation Contract: inputs/intermediate state/outputs explicit; no
hidden global state, implicit random choices, or silent downstream
re-decisions. Determinism: reproducible from explicit seeds/versions.
"""

from __future__ import annotations
from typing import List, Optional
from .datamodel import FinalValidatedPerformance, PerformanceEvent


class PerformanceRuntime:
    """RULE 26 primary object. Purely a *reader/player* of an already-solved
    FinalValidatedPerformance — it never calls back into Rules 1-25/29-40 to
    re-decide anything, satisfying the 'no hidden re-sticking' principle."""

    def __init__(self, performance: FinalValidatedPerformance):
        if not performance.validation.approved:
            raise ValueError("cannot construct a runtime from an unapproved performance (Rule 25 gate)")
        self.performance = performance
        self.playhead_s: float = 0.0
        self.playing: bool = False
        self._events_sorted: List[PerformanceEvent] = sorted(
            performance.events, key=lambda e: e.time_seconds
        )

    def play(self) -> None:
        self.playing = True

    def pause(self) -> None:
        self.playing = False

    def seek(self, time_s: float) -> None:
        self.playhead_s = max(0.0, min(self.performance.duration_s, time_s))

    def events_between(self, t0: float, t1: float) -> List[PerformanceEvent]:
        """Deterministic query used by a real-time host to pull exactly the
        events due in a frame window — same seed/version always returns the
        same slice."""
        return [e for e in self._events_sorted if t0 <= e.time_seconds < t1]

    def advance(self, dt_s: float) -> List[PerformanceEvent]:
        if not self.playing:
            return []
        t0 = self.playhead_s
        t1 = min(self.performance.duration_s, self.playhead_s + dt_s)
        due = self.events_between(t0, t1)
        self.playhead_s = t1
        return due

    def export_summary(self) -> dict:
        return {
            "engine_version": self.performance.engine_version,
            "seed": self.performance.seed,
            "duration_s": self.performance.duration_s,
            "event_count": len(self.performance.events),
            "validation_issues": len(self.performance.validation.issues),
        }
